src/cargo/ops/registry/publish.rs

```diff
@@ -115,82 +123,165 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+                };
+                registry.set_token(Some(auth::auth_token(
+                    &opts.gctx,
+                    &source_ids.original,
+                    None,
+                    operation,
+                    vec![],
+                    false,
+                )?));
+            }
 
-    opts.gctx
-        .shell()
-        .status("Uploading", pkg.package_id().to_string())?;
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+            transmit(
+                opts.gctx,
+                ws,
+                pkg,
+                tarball.file(),
+                &mut registry,
+                source_ids.original,
+                opts.dry_run,
+            )?;
+
+            if !opts.dry_run {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
+            }
+        }
+
+        to_confirm.extend(ready.iter().copied());
+        let confirmed = if opts.dry_run {
+            to_confirm.clone()
         } else {
-            DEFAULT_TIMEOUT
+            const DEFAULT_TIMEOUT: u64 = 60;
+            let timeout = if opts.gctx.cli_unstable().publish_timeout {
+                let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
+                timeout.unwrap_or(DEFAULT_TIMEOUT)
+            } else {
+                DEFAULT_TIMEOUT
+            };
+            if 0 < timeout {
+                // Note that we're applying the timeout on every pass through
+                // the loop, so that if there's a big dependency graph then the
+                // "effective" timeout could be much larger. Maybe this is not
+                // the right behavior.
+                let timeout = Duration::from_secs(timeout);
+                wait_for_any_publish_confirmation(
+                    opts.gctx,
+                    source_ids.original,
+                    &to_confirm,
+                    timeout,
+                )?
+            } else {
+                BTreeSet::new()
+            }
         };
-        if 0 < timeout {
-            let timeout = Duration::from_secs(timeout);
-            wait_for_publish(opts.gctx, source_ids.original, pkg, timeout)?;
+
+        for id in &confirmed {
+            to_confirm.remove(id);
+            remaining.remove(id);
```

Comment:
What if we removed from `remaining` when we transmit?  The loop will need to be updated to be `while !remaining.is_empty() || !to_confirm.is_empty()` which might negate the need for the comment I suggested.

Wait, do we need `remaining`?  If `PublishOrder` removes 0 weights, we can have a `any_ready` function that checks if weights is 0.  By removing `remaining`, we reduce the number of data structure we need to keep in sync and remove the question of "remaining to do what" (I keep assuming "to upload" while currently its" to do anything").